File: src/main/java/jerre/kotlin/workshop/models/Address.java
package jerre.kotlin.workshop.models;

import java.util.Objects;
import java.util.Optional;

public class Address {
    private String streetName;
    private Integer streetNumber;

    public Address(String streetName, Integer streetNumber) {
        this.streetName = streetName;
        this.streetNumber = streetNumber;
    }

    public Address(String streetName) {
        this(streetName, null);
    }

    public String getStreetName() {
        return streetName;
    }

    public void setStreetName(String streetName) {
        this.streetName = streetName;
    }

    public Optional<Integer> getStreetNumber() {
        return Optional.ofNullable(streetNumber);
    }

    public void setStreetNumber(Integer streetNumber) {
        this.streetNumber = streetNumber;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Address address = (Address) o;
        return Objects.equals(streetName, address.streetName) &&
                Objects.equals(streetNumber, address.streetNumber);
    }

    @Override
    public int hashCode() {
        return Objects.hash(streetName, streetNumber);
    }

    @Override
    public String toString() {
        return "Address{" +
                "streetName='" + streetName + '\'' +
                ", streetNumber=" + streetNumber +
                '}';
    }

    public String pretty() {
        String prettyStreet = streetName == null ? "Unknown streetName" : streetName;
        String prettyStreetNumber = this.streetNumber == null ? "" : this.streetNumber.toString();
        return String.format("%s %s", prettyStreet, prettyStreetNumber);
    }
}

File: src/main/kotlin/jerre/models/Address.kt
package jerre.models

import java.util.*

class AddressJalla(streetNameParam: String, streetNumberParam: Int?) {
    constructor(streetName: String): this(streetNameParam = streetName, streetNumberParam = null)

    private val streetName: String
    private val streetNumber: Int?

    init {
        this.streetName = streetNameParam
        this.streetNumber = streetNumberParam
    }

    fun getStreetName(): String {
        return this.streetName
    }

    fun getStreetNumber(): Int? {
        return this.streetNumber
    }

    override fun equals(other: Any?): Boolean {
        if (this == other) return true
        if (other == null || this.javaClass != other.javaClass) return false
        val address = other as AddressJalla
        return Objects.equals(streetName, address.streetName)
                && Objects.equals(streetNumber, address.streetNumber)
    }

    override fun hashCode(): Int {
        return Objects.hash(streetNumber, streetNumber)
    }

    override fun toString(): String {
        return "Address{" +
                "streetName='" + streetName + "\'" +
                ", streetNumber=" + streetNumber +
                "}"
    }

    fun pretty(): String {
        val prettyStreetName = streetName   //Trenger ikke null-sjekk her siden streetName ikke kan være null
        val prettyStreetNumber = if (streetNumber == null) "" else streetNumber
        return streetName + " " + streetNumber
    }
}

class AddressMindreJalla(val streetName: String, val streetNumber: Int? = null) {

    override fun equals(other: Any?): Boolean {
        if (this == other) return true
        if (other == null || this.javaClass != other.javaClass) return false
        val address = other as AddressJalla
        return Objects.equals(streetName, address.getStreetName())
                && Objects.equals(streetNumber, address.getStreetName())
    }

    override fun hashCode(): Int {
        return Objects.hash(streetNumber, streetNumber)
    }

    override fun toString(): String {
        return "Address{" +
                "streetName='" + streetName + "\'" +
                ", streetNumber=" + streetNumber +
                "}"
    }

    fun pretty(): String {
        val prettyStreetName = streetName   //Trenger ikke null-sjekk her siden streetName ikke kan være null
        val prettyStreetNumber = if (streetNumber == null) "" else streetNumber
        return streetName + " " + streetNumber
    }
}

data class Address @JvmOverloads constructor(val streetName: String, val streetNumber: Int? = null) {
    fun pretty(): String = "$streetName ${streetNumber ?: ""}"
}

fun main(args: Array<String>) {
    //Jalla varianten
    var addressJalla = AddressJalla(streetNameParam = "Gata", streetNumberParam = 1)
//    var addressJalla = AddressJalla("Gata")
    println(addressJalla)
    println("streetName: ${addressJalla.getStreetName()}")
    println("streetNumber: ${addressJalla.getStreetNumber()}")
    println("pretty: ${addressJalla.pretty()}")
    println()

    //Mindre jalla variant
    var addressMindreJalla = AddressMindreJalla(streetNumber = 1, streetName = "Gata")
//    var addressMindreJalla = AddressMindreJalla("Gata")
    println(addressMindreJalla)
    println("streetName: ${addressMindreJalla.streetName}")
    println("streetNumber: ${addressMindreJalla.streetNumber}")
    println("pretty: ${addressMindreJalla.pretty()}")
    println()

    //Kotlin variant
    var address = Address(streetName = "Gata", streetNumber = 1)
//    var address = Address("Gata")

    println(addressMindreJalla)
    println("streetName: ${addressMindreJalla.streetName}")
    println("streetNumber: ${addressMindreJalla.streetNumber}")
    println("pretty: ${addressMindreJalla.pretty()}")

    var addressCopy = address.copy()
    println("copy is same: ${address == addressCopy}")

    var neighbour = address.copy(streetNumber = 2)
    println("neighbour address: $neighbour")
}